rust/arrow/src/compute/kernels/sort.rs

```diff
@@ -283,6 +285,30 @@ mod tests {
             None,
             vec![3, 1, 4, 2, 0, 5],
         );
+        test_sort_to_indices_primitive_arrays::<Float32Type>(
+            vec![
+                None,
+                Some(-0.05),
+                Some(2.225),
+                Some(-1.01),
+                Some(-0.05),
+                None,
+            ],
+            None,
+            vec![3, 1, 4, 2, 0, 5],
+        );
+        test_sort_to_indices_primitive_arrays::<Float64Type>(
+            vec![
+                None,
+                Some(-0.05),
+                Some(2.225),
+                Some(-1.01),
+                Some(-0.05),
+                None,
```

Comment:
Could you add `f64::NAN` to the test values.

Reply:
I see that we do get the failure if we have a `f{32|64}::NAN`. Handling `NAN` might complicate the integer case, so should I separate the float implementation?

How do we treat NaN in general across the Rust implementation?

Reply:
I can look at this in more detail over the weekend but I think having special impl for f32/f64 is probably the way to go, then you can add specific checks for `f32::NAN` and `f64::NAN` before calling `partial_cmp` and we'll need to decide if `NaN` comes before or after valid numbers.

Reply:
PTAL at the solution that I've just pushed. Given that the `unwrap()` will only fail if there's a `NAN`, I've replaced it by a default `std::cmp::Ordering::Greater` to treat `NAN` as the highest value. In the descending sort path, it ends up being inverted to the lowest value.

I suppose a better approach might be to let the `nulls_first` sort option drive the behaviour, with the below sort options:

- ascending, nulls last: `Ordering::Greater` placing NaNs before the first null 
- ascending, nulls first: `Ordering::Less` placing NaNs after the last null
- descending, nulls last: `Ordering::Greater`
- descending, nulls first: `Ordering::Less`

... so the ordering being determined by the null behaviour (it helped to write it out :smile:)

Reply:
The problem with this approach is that the NaN could be a, or b, or both so this comparison is now non-deterministic and inconsistent. I think implementing this specifically for Float32Array and Float64Array and checking for NaN on both values is the only way we can handle this correctly.

Reply:
PTAL at my changes, I check for NaN values when I perform the range partition, so that there are no nulls when values are sorted.